pkg/apis/cluster/v1alpha1/common_types.go

```diff
@@ -56,3 +57,69 @@ type MachineClassRef struct {
 	// +optional
 	Provider string `json:"provider,omitempty"`
 }
+
+// ObjectMeta is metadata that all persisted resources must have, which includes all objects
```

Comment:
Can we leave more context? For example, why is this necessary for `Machine` and not other CRDs? And if it's only because `Machine` is embedded in other CRDs, then is there an upstream ticket to track this issue?

Reply:
Added more context

Reply:
This is a fair explanation. I was mostly thinking about the problem this attempts to solve, and how it is likely to come up again. For example, suppose someone embeds `MachineDeployment` in another CRD. Won't they run into these same issues? And if so, why not remove `ObjectMeta` from `MachineDeployment` too?

I guess I am wondering what the long term guidance is from `apimachinery`...

Reply:
There's a couple of things going on here:

1. All objects have root-level `ObjectMeta`
1. Some objects, such as `Machine` and `MachineSet`, include non-root `ObjectMeta` somewhere in their nested structures

If anyone includes any Kubernetes object (built-in or CRD) in their own object, you'll end up with a similar situation: root-level `ObjectMetadata` along with non-root `ObjectMeta`. In the past, we've tried to design the API types such that the portions that someone might want to include are reusable (e.g. including `MachineSpec` in `MachineSet` (instead of a full `Machine`)). This could work, in theory, except some of these types have an `ObjectMeta` in them, even though they're not root-level types. It would be nice if these types used some other structure instead of `ObjectMeta`.

That's not necessarily a problem - we've been doing it for a while in CAPI, and core Kubernetes does it too. The problem @vincepri is running into is that the newer version of kubebuilder/controller-runtime/controller-tools generates openapi schema validation definitions that make it so you can't successfully create a Machine because of this nested `ObjectMeta` and a null `creationTimestamp` field.

Reply:
Also, when including a non-root `ObjectMeta` field, a lot of the time it's just to use it for the `annotations` and/or `labels`, but not the entire set of fields. It may be better to scope the non-root field to just the fields you need, instead of a full `ObjectMeta` struct.

Reply:
+1 on inspecting long term effects.

i think the PR should include more detail on the actual problem it's solving.
e.g. comments about:
```
validation failure list:
            spec.metadata.creationTimestamp in body must be of type string: "null"
```


something puzzling in the slack thread was that this started happening suddenly, did you figure out what caused that? was it https://github.com/kubernetes/kubernetes/pull/77653 ?